DFC.TestAutomation.UI/Helper/CommonActionHelper.cs

```diff
@@ -0,0 +1,239 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace DFC.TestAutomation.UI.Helper
+{
+    /// <summary>
+    /// Provides helper functions for all common operations.
+    /// </summary>
+    public class CommonActionHelper : ICommonActionHelper
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CommonActionHelper"/> class.
+        /// </summary>
+        /// <param name="webDriver">The Selenium webdriver.</param>
+        /// <param name="webDriverWaitHelper">The Selenium webdriver wait helper.</param>
+        /// <param name="retryHelper">The retry helper.</param>
+        public CommonActionHelper(IWebDriver webDriver, IWebDriverWaitHelper webDriverWaitHelper, IRetryHelper retryHelper)
+        {
+            this.WebDriver = webDriver;
+            this.WebDriverWaitHelper = webDriverWaitHelper;
+            this.RetryHelper = retryHelper;
+        }
+
+        private IWebDriver WebDriver { get; set; }
+
+        private IWebDriverWaitHelper WebDriverWaitHelper { get; set; }
+
+        private IRetryHelper RetryHelper { get; set; }
+
+        /// <summary>
+        /// Assesses whether the IWebElement contains expected text. The comparison is case insensitive and the expected text parameter
+        /// is trimmed before being compared.
+        /// </summary>
+        /// <param name="locator">The IWebElement locator.</param>
+        /// <param name="expectedText">The expected text.</param>
+        /// <returns>True if the IWebElement contains the expected text. False if the IWebElement does not contain the expected text.</returns>
+        public bool ElementContainsText(By locator, string expectedText)
+        {
+            bool Action()
+            {
+                this.WebDriverWaitHelper.WaitForPageToLoad();
+                var actual = this.GetText(locator);
+                return actual.Contains(expectedText.Trim(), StringComparison.CurrentCultureIgnoreCase);
+            }
+
+            return this.RetryHelper.RetryOnException(Action);
+        }
+
+        /// <summary>
+        /// Gets the attribute value for an IWebElement.
+        /// </summary>
+        /// <param name="locator">The IWebElement locator.</param>
+        /// <param name="attributeName">The attribute name.</param>
+        /// <returns>The atrribute value.</returns>
+        public string GetAttributeValue(By locator, string attributeName)
+        {
+            string Action()
+            {
+                return this.WebDriver.FindElement(locator).GetAttribute(attributeName);
+            }
+
+            return this.RetryHelper.RetryOnException<string>(Action);
+        }
+
+        /// <summary>
+        /// Gets the text value from all IWebElements with a given locator.
+        /// </summary>
+        /// <param name="locator">The IWebELement locator.</param>
+        /// <returns>The string value.</returns>
+        public string GetTextFromElements(By locator)
```

Comment:
Does this return text in a suitable structure for use? e.g. how would a request for text from a table be returned? ...or within a <DIV> - would paragraphs be conjoined in a single concatenated string?

Reply:
It simply returns the text value displayed. I understand your reasons for asking this but I feel like this could be a very deep hole to start digging. We can 'clean up' text but to what end? I'd be more inclined to have string extension methods that handle newline characters or different encoding types and allow the user to do what they want with the text value rather than assuming they want it 'cleaned up'.

I think this goes back to the approach - do it when it's needed. Lets see how we get on and also lets remember that this package is public. So it ought to work across a variety of web applications, not just for the NCS.